### ISSUE TITLE
Typesense is failing to recognize "space" as a typo.

### ISSUE THREAD

**Issue by sushant4849** [typesense/typesense#133]
Nice project. But the default behaviour is totally failing with compound words. 

For example, I've indexed a document titled "basketball scores". If the search query is "basketball" or "basketball scores" or even "baketball scores", it is perfectly retrieving the document.

But if the search query is "basket ball", it is bringing up totally irrelevant documents to the top. In the minds of the user "basketball scores" & "basket ball scores" are exactly the same query.

Similarly I tested with other compound words like "bedroom", "fingerprints", "DaVinciCode etc and it fails if the query is "bed room", "finger prints", "Da vinci code".

Is there anyway to fix this default behaviour and pass space as a potential typo.

**Comment by kishorenc** [comment]
It's not possible at the moment to treat space as a typo but I can see how this can be a useful feature to have.

**Comment by simonsinclair** [comment]
@kishorenc is there anything that can be done to work around this?

I have a foods database and would like user queries for "chickpea" to return documents matching "chick pea". I think other search engines call this behaviour Phrase Queries.

**Comment by kishorenc** [comment]
@simonsinclair I've prioritized this feature, so this will be addressed really soon.

**Comment by simonsinclair** [comment]
That's super news, @kishorenc. Thank you!

**Comment by kishorenc** [comment]
Looking into this in more detail. Let's say there is a query containing 3 tokens: "alpha beta gamma". 

Currently, if there are no documents that contain all 3 tokens exactly, Typesense does a query "rewrite" in the following order:

1. Look for tokens with typos that match closely to the tokens in the query and try to fetch documents matching those "corrected" tokens. Tokens with upto 2 typos are considered (controlled by the `num_typos` parameter).
2. If we still don't find any hits, we try to drop tokens in the query and see if there are documents that contain 2 of the 3 tokens or just one of the tokens (in that order).

For compounding, I'm thinking of of rewriting the query with variations containing compound words, i.e. `alphabetagamma`, `alphabeta gamma`, `alpha betagama` etc. before attempting for typos. The number of combinations involved will increase as the number of tokens in the query increases: roughly `nC2+nC3+...+nCn` where `n` is the number of tokens in the query. There are a few optimizations we can do to reduce that complexity: 

1. do only 2 or 3 token compounds (so we will not do a 4-token compound like `alphabetagammaomega`)
2. do only compounds for queries up to a given length (say upto 5 token queries)

Approach 1 is probably more deterministic. There is also the case of the original query containing a typo, thereby requiring both compounding and typo correction, e.g. `chhck pea` -> `chickpea`.

Finally, similar to compounding the tokens in a query to search for results , we will also have to handle the possibility of having to break a query apart, e.g. query contains `chickpea` but documents only contain `chick pea`. This will require finding the first character in the `chickpea` sequence that produces a new word so we can break the single token apart and query for resulting individual tokens. The same caveats around complexity holds.

**Comment by simonsinclair** [comment]
Thank you for explaining, @kishorenc. So, if I understand correctly, the proposed order of operations is:

1. Rewrite the search query with variations containing compound words to correct typos.
    - I support the more deterministic approach of optimising with solution 1.

Then, if no hits:

2. Look for tokens that closely match tokens in the search query to correct typos. Then, if no hits:
3. Begin to drop tokens to broaden the search query.

Where does the second case (breaking a search query apart) fit into this? I imagine correcting typos or broadening the search query _after_ splitting compound words could produce irrelevant results? We know that tokens created by splitting compound words should always be next to each other. E.G. `newspaper`, `news paper` and not `news` or `new`.

Out of curiosity, do any of these operations ever happen in parallel? If they were to both produce results, could they be combined?

I don't have a lot of experience with this kind of processing, but it is interesting! Thanks again.

**Comment by kishorenc** [comment]
This is fixed in `v0.23`.

**Comment by entrptaher** [comment]
@kishorenc I am testing it on v0.23, kindly let me know if this is similar. I have **sol punks**

![image](https://user-images.githubusercontent.com/8284972/175784583-5ede96eb-b888-496f-a60d-d6b324bba055.png)

But what I also want to include in the search result is this one without space.

![image](https://user-images.githubusercontent.com/8284972/175784596-14453908-47a7-40cd-a909-1a1b980e96af.png)

If I try any other combination like this, it works, 

![image](https://user-images.githubusercontent.com/8284972/175784643-429de731-c8bc-4468-8e4c-730d79b7caf1.png)

Basically everything works except **sol punks** with a space.
